teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,203 @@
+      name: task-run-example
+    retries: 1
+    timeout: "10s"
+  - name: custom-run-example
+    taskRef:
+      apiVersion: example.dev/v1alpha1
+      kind: Example
+    retries: 1
+    timeout: "10s"
+```
+
+Customers define a TaskRun and a CustomRun within a Pipeline. Even though they set up the same value for the same fields (`timeout` and `retries`), the two runtime objects (`task-run-example` and `custom-run-example`) would behave differently. One (`task-run-example`) will retry once after 10s, while the other (`custom-run-example`) will be cancelled.
+
+#### Retry TaskRun Independently
+
+As a standalone runtime object, TaskRuns can be used independently (outside of a PipelineRun) in production environment, here are two examples:
+- https://github.com/tektoncd/catalog/tree/main/task/send-to-webhook-slack/0.1
+- https://github.com/tektoncd/catalog/tree/main/task/sendmail/0.1
+
+Transient errors are everywhere especially in the Cloud Environment, services can be down for a short period of time making the entire TaskRun fails. https://learn.microsoft.com/en-us/azure/architecture/best-practices/transient-faults#why-do-transient-faults-occur-in-the-cloud explains how common the transient errors are in the Cloud env. 
+
+With retries supported, customers are able to write robust TaskRuns to support such use cases.
+
+## Related Work
+
+In the CI/CD industry, most systems support retry, some of them enables users to configure the retry strategy[^retry-strategy].
+
+Typically, a retry strategy needs to specify:
+1. When to retry
+2. The amount of attempts
+3. Actions to take after a failed attempt
+4. Timeout of each attempt
+
+
+|  | [Retry Step in GA](https://github.com/marketplace/actions/retry-step) | [Retry Action in GA](https://github.com/marketplace/actions/retry-action) | [Argo Workflow](https://argoproj.github.io/argo-workflows/retries/#retry-policies) | [GitLab](https://docs.gitlab.com/ee/ci/yaml/#retry) | [Ansible](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html#retrying-a-task-until-a-condition-is-met)| [Concourse](https://concourse-ci.org/attempts-step.html#attempts-step) |
+|:---|:---|:---|:---|:---|:---|:---|
+| **When to Retry** | configurable | on failure |configurable|configurable|[always retry, conditional stop](https://github.com/ansible/ansible/pull/76101) [^ansible-conditional-stop]|configurable|
+| **Attempts amount** |supported|supported|supported|supported|supported|supported|
+| **Timeout for each attempt** |supported|supported|supported|[supported](https://docs.gitlab.com/ee/ci/yaml/#retrywhen)|supported|supported|
+| **Timeout for all attempts** |[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[?](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout)|-|-|-|
+
```

Comment:
Thank you @XinruZhang for this detailed comparison with an industry standard.

Tekton pipelines do support [retries](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#using-the-retries-field) when a task is part of a pipeline. The task can be a native Tekton task or a custom task.

This table does not represent apples to apples comparison. For example, Argo Workflow is not equivalent to Tekton `task` but its more or less similar to `pipeline` or even Tekton `workflow`. This proposal is focused on independent `taskRun` and `customRun` CRDs. We need to identify and document the list of resources in the industry which are equivalent to a `task` or `taskRun` in Tekton.

This table is misleading and may confuse users while comparing Tekton with the other CI/CD systems.

@jerop and I had documented this kind of comparison last year (which might be outdated at this point), but something for your reference:

![Screen Shot 2022-09-28 at 10 15 36 AM](https://user-images.githubusercontent.com/206285/192846900-4dbcd083-e78b-4c74-ac55-2debc902ee74.png)

Reply:
Thanks for the review @pritidesai and the comparison work you and Jerop did!

This related work section is written for comparing the general retry strategy (instead of demonstrating that retry is a neccesary functionality in TaskRun), especially about if we should set Timeout for all retry attempts or each retry attempt. I updated the Related Work section, hopefuly the confusion is alleviated.

Thanks for the concept comparison! It seems like there are no concepts equivalent to Task, becuase in Tekton, PipelineRun, TaskRun and CustomRun can all run on their own, while in other CI/CD systems, they only support the top level Pipeline/Workflow to run on its own. So it's hard for us to find an industry standard to follow to decide if we want to support retry in TaskRun.